c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -295,18 +301,39 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                     Some(Callee::MiscBuiltin) => {}
 
                     Some(Callee::Malloc) => {
-                        self.visit_place(destination, Mutability::Mut);
+                        let destination = self
+                            .acx
+                            .special_casts
+                            .get(&destination)
+                            .unwrap_or(&destination);
+                        self.visit_place(*destination, Mutability::Mut);
                     }
 
                     Some(Callee::Calloc) => {
-                        self.visit_place(destination, Mutability::Mut);
+                        let destination = self
+                            .acx
+                            .special_casts
+                            .get(&destination)
+                            .unwrap_or(&destination);
```

Comment:
This logic of `special_casts.get(&x).unwrap_or(&x)` that's repeated a bunch of times could be put in `SpecialCasts`, too.